### PR TITLE
[Impeller] flood input coverage with destructive color filter.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/126389
Fixes https://github.com/flutter/flutter/issues/137090
Fixes https://github.com/flutter/flutter/issues/154035

If this bit is set, the output of a color filter should flood to the input of a filter to parent clip. We're not checking this right now and so all color filter content is treated as bounded.

This needs to flood the input as the image filter (which occurs afterwards) should process the flooded input.

### CODE REVIEW

**Review by bdero — APPROVED** [review]
LGTM!

**Comment by bdero** [comment]
Hmm, actually I think something is going awry with `TranslucentSaveLayerWithColorAndImageFilterDrawsCorrectly`.

The red source color on the SaveLayer paint's color filter is leaking through even through the blend mode is set to `kDstOver`.

**Comment by jonahwilliams** [comment]
https://github.com/flutter/engine/blob/main/display_list/effects/dl_color_filter.cc#L55 The Dl ColorFilter marks dstOver with a non-transparent color as modifying transparent black

FYI @flar

**Comment by flar** [comment]
> https://github.com/flutter/engine/blob/main/display_list/effects/dl_color_filter.cc#L55 The Dl ColorFilter marks dstOver with a non-transparent color as modifying transparent black
> 
> FYI @flar

Isn't that correct?

Dst is transparent, color is not. The result with dstOver would be the color which means it modifies transparent black...?

See https://github.com/flutter/engine/blob/066a3c71187bb3b6ef694decb9c32287bcd06e45/display_list/testing/dl_rendering_unittests.cc#L4297

**Comment by bdero** [comment]
Oh, this _is_ actually correct behavior. @flar is right. Because most of the destination is transparent, the red source color _should_ be leaking through everywhere except for the partially translucent black rectangle.

So this LGTM.

**Comment by flar** [comment]
> Oh, this _is_ actually correct behavior. @flar is right. I confused `kDst` and `kDstOver`. Because most of the destination is transparent, the red source color _should_ be leaking through everywhere except for the partially translucent black rectangle.
> 
> So this LGTM.

But we should see if the sudden change in behavior is masking what the test was originally trying to test. In particular the test mentioned above is missing the blue rectangle now. Was the blue rectangle important? If so, then let's adjust the test to match its own purpose...

**Comment by bdero** [comment]
- For `TranslucentSaveLayerWithBlendColorFilterDrawsCorrectly`, `TranslucentSaveLayerWithColorAndImageFilterDrawsCorrectly` I think the value that the blue rectangle adds is a visual comparison with a filter-less `SaveLayer`, so I think just swapping the order around to make the blue rectangle appear over the flooded red area would keep that value.
- `ForegroundBlendSubpassCollapseOptimization` looks good as is. The before image just looks incorrect.
- For `BlendModePlusAlphaColorFilterWideGamut`, the dark green background is being blended with the SaveLayer drawn over it with `kPlus`. I don't think exposing the dark green background adds any value, so it seems fine as-is.

But if we want to prevent any or all of these goldens from changing, it can be done by passing bounds rects to the SaveLayers of course.

**Comment by jonahwilliams** [comment]
I updated TranslucentSaveLayerWithBlendColorFilterDrawsCorrectly to clip so it matches the old golden, for the rest I'll update.

File: impeller/display_list/aiks_dl_unittests.cc [hunk 1]
```diff
@@ -160,12 +160,15 @@ TEST_P(AiksTest, TranslucentSaveLayerWithBlendColorFilterDrawsCorrectly) {
   paint.setColor(DlColor::kBlack().withAlpha(128));
   paint.setColorFilter(
       DlBlendColorFilter::Make(DlColor::kRed(), DlBlendMode::kDstOver));
+  builder.Save();
+  builder.ClipRect(SkRect::MakeXYWH(100, 500, 300, 300));
```
Comment (bdero) [inline impeller/display_list/aiks_dl_unittests.cc:164]: Should we update `TranslucentSaveLayerWithColorAndImageFilterDrawsCorrectly` too?

Comment (jonahwilliams) [reply impeller/display_list/aiks_dl_unittests.cc:164]: Done

**Comment by flar** [comment]
> * For `TranslucentSaveLayerWithBlendColorFilterDrawsCorrectly`, `TranslucentSaveLayerWithColorAndImageFilterDrawsCorrectly` I think the value that the blue rectangle adds is a visual comparison with a filter-less `SaveLayer`, so I think just swapping the order around to make the blue rectangle appear over the flooded red area would keep that value.
> * `ForegroundBlendSubpassCollapseOptimization` looks good as is. The before image just looks incorrect.
> * For `BlendModePlusAlphaColorFilterWideGamut`, the dark green background is being blended with the SaveLayer drawn over it with `kPlus`. I don't think exposing the dark green background adds any value, so it seems fine as-is.
> 
> But if we want to prevent any or all of these goldens from changing, it can be done by passing bounds rects to the SaveLayers of course.

I was going to say "add a clip" which it looks like Jonah did. If no changes to the goldens are the biggest concern then clipping to the SL region would do that. If you want to also have this test show that it overflows (the new behavior) then the clip could be slightly bigger so that it both shows all the old content and also shows that the overflow is happening without obscuring the other part of the test.

File: impeller/display_list/aiks_dl_unittests.cc (hunk 1 above)
Comment (flar) [reply impeller/display_list/aiks_dl_unittests.cc:164]: The test name has "Translucent SaveLayer" in it which makes me think that the expanded flood is part of the behavior it is wanting to demonstrate. By clipping so tightly there is no indication of the flooding when such a combination happens. Perhaps expand the clip by 10 pixels or so, which will both show the flooding without completely obscuring the blue rectangle?

**Comment by jonahwilliams** [comment]
I did add some new test goldens that show the flooding,  that is DestructiveBlendColorFilterFloodsClip

**Pull request merged**